pkg/reconciler/route/domains/domains.go

```diff
@@ -0,0 +1,51 @@
+package domains
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/serving/pkg/network"
+	"github.com/knative/serving/pkg/reconciler/route/config"
+)
+
+const (
+	// HTTPScheme is the string representation of http.
+	HTTPScheme string = "http"
+)
+
+// SubdomainName generates a name which represents the subdomain of a route
+func SubdomainName(r *v1alpha1.Route, suffix string) string {
+	if suffix == "" {
+		return r.Name
+	}
+	return fmt.Sprintf("%s-%s", r.Name, suffix)
+}
+
+// DomainNameFromTemplate generates domain name base on the template specified in the `config-network` ConfigMap.
+// name is the "subdomain" which will be referred as the "name" in the template
+func DomainNameFromTemplate(ctx context.Context, r *v1alpha1.Route, name string) (string, error) {
+	domainConfig := config.FromContext(ctx).Domain
+	domain := domainConfig.LookupDomainForLabels(r.ObjectMeta.Labels)
+
+	// These are the available properties they can choose from.
+	// We could add more over time - e.g. RevisionName if we thought that
+	// might be of interest to people.
+	data := network.DomainTemplateValues{
+		Name:      name,
+		Namespace: r.Namespace,
+		Domain:    domain,
+	}
+
+	networkConfig := config.FromContext(ctx).Network
+	buf := bytes.Buffer{}
+	if err := networkConfig.GetDomainTemplate().Execute(&buf, data); err != nil {
+		return "", fmt.Errorf("error executing the DomainTemplate: %s", err)
+	}
+	return buf.String(), nil
+}
+
+// URL generates the a string representation of a URL.
+func URL(scheme, fqdn string) string {
+	return fmt.Sprintf("%s://%s", scheme, fqdn)
```

Comment:
May be do it like this?
https://play.golang.org/p/9BFnU3nWCMc

Reply:
That suggestion works. (This was a cut & paste from traffic.go)